DublinCore and ArXiv/DC-OAIPMH parsers added

seasidesparrow opened the pull request:
Reissued pull request to cover completed version of parser and run.py, along with preliminary test suite.

seasidesparrow commented:
I am going to push a new version of this that separates out the requirements and code for the direct ingest pipeline.  The file serializers/serializer_mp.py will live with the direct_ingest repo for now.  Everything else should be stand-alone.

aaccomazzi commented on pyingest/parsers/arxiv.py:
```diff
@@ -0,0 +1,133 @@
+    if len(arxiv_id2) == 4:
+        arxiv_id2 = u'.'+arxiv_id2
+
+    auth_init = authors[0][0]
```
Need to consider cases of:
1. no author (gets a '.')
2. Unicode character (needs ascii downgrade)

aaccomazzi commented on pyingest/parsers/dubcore.py:
```diff
@@ -0,0 +1,87 @@
+    def get_tag(self, r, tag):
+        value = []
+        try:
+            tag,self._array(r.get(tag,[]))
```
Don't quite understand why this is necessary (calling the self._array function twice, once inside of try and then outside).
Also r.get(key,default) should never generate a KeyError AFAIK

aaccomazzi commented on test/test_parsers.py:
```diff
@@ -47,4 +48,35 @@ def test_zenodo_parser(self):
 
 
 
+class TestArxiv(unittest.TestCase):
```
I suggest adding test cases for arXiv records where the first author's initial is a non-ascii character.

seasidesparrow commented on pyingest/parsers/arxiv.py:
```diff
@@ -0,0 +1,133 @@
+        else:
+            testdoi,url = make_extras(r['dc:identifier'])
+            if testdoi != [None]:
+                arx['doi'] = testdoi
```
My logic was that I didn't want the kw pair arx['doi'] to exist (i.e. undefined, and not None) if there were no DOI present.

seasidesparrow commented on pyingest/parsers/arxiv.py:
```diff
@@ -0,0 +1,133 @@
+            raise EmptyParserException("No dictionary.")
+
+        try:
+            arx['pubdate']  = r['dc:date'][-1]
```
There are sometimes multiple dates, and I think that's how their version history is handled.  IIRC, this value always returns an array, even if it's only one date, and the most recent date is at the end of the array.  For new entries, date[0] = date[-1]; for revisions, date[-1] is the current (most recent) revision.

seasidesparrow commented on pyingest/parsers/arxiv.py:
```diff
@@ -0,0 +1,133 @@
+    return [doi],url
+
+
+def make_bibcode(url,authors):
```
Paragraph 1: Yes I did.  I seem to remember encountering some error at some point doing that.  But yes, I agree with what you're saying, and I will try to track down and handle whatever error I got using the ID.

Paragraph 2: This is a good point, and it's why I'd like to meet on Monday.  When we last talked about the specs for this, I got the impression that the bibcode didn't have to rigorously match what the classic parser gave because they'd be reconciled.  But in light of discussion at the Jan 26 scrum, I wonder if this is not true?  Regardless, I will implement the existing bibcode generation as exactly as I can if it is needed.

seasidesparrow commented:
A followup to a separate conversation we had: I've removed the xmltodict library from the code, and added the package as a requirement.  Currently it's pointing to xmltodict==0.9.2, but I don't think it would be a problem to use current.

seasidesparrow commented:
Hi Alberto,

The arXiv parser portion of this is done.  I think I've addressed everything you asked for, either rewriting things or removing them.  Also added a test case to check the bibcodes for several old-style arxiv subject-heading/NNNNNNN papers.

aaccomazzi commented on pyingest/parsers/arxiv.py:
```diff
@@ -0,0 +1,129 @@
+                    arx_num = arx_num[2:]
+            else:
+                arx_field = 'arXiv'
+                arx_num = arxiv_id.replace('.','')
```
I think this does not do the right thing for the "4.4" format of arXiv records, eg. arXiv:0704.0010, please double check

aaccomazzi commented on pyingest/tests/test_parsers.py:
```diff
@@ -0,0 +1,85 @@
+        self.assertNotEqual(shouldbe['title'],document['title'])
+
+    def test_unicode_init(self):
+        shouldbe = {'bibcode':u'2009arXiv.09012443O'}
```
Real bibcode is 2009arXiv0901.2443O

aaccomazzi commented on pyingest/parsers/dubcore.py:
```diff
@@ -0,0 +1,156 @@
+#abstract page
+
+            if self.get_tag(r,'dc:identifier'):
+                output_metadata['properties'] = self.get_tag(r,'dc:identifier')
```
I don't think this does what we want.  The 'properties' element should be a dict so that we can keep things such as links to fulltext, DOIs, etc.  See e.g. this example from the zenodo test files:
```
    "properties": {
        "DOI": "10.5281/zenodo.16803", 
        "ELECTR": "http://zenodo.org/record/16803", 
        "OPEN": 1, 
        "PDF": "https://zenodo.org/record/16803/files/Murphy_Thesis.pdf"
    }, 
```

aaccomazzi commented on pyingest/parsers/dubcore.py:
```diff
@@ -0,0 +1,156 @@
+            self.check_schema(r)
+        except:
+            raise UnparseableException("Cannot parse record.")
+        else:
```
Just a style comment: AFAIK there is no need to have the code below be in the `else` statement.  Once we throw an exception, we are guaranteed that no remaining code gets executed below.

seasidesparrow commented:
I've updated everything except the style correction to dubcore.py -- I'd like to read what's standard in PEP-8, but I like what I have in the sense that it makes the program flow more obvious to me.  

I've put serialized tagged versions of the test oai-pmh files into test_files/arxiv.test/tagged, but since they were generated with the same code I'm testing, I need to validate them.
